In [76]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report


---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [77]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
crdf=pd.read_csv('lending_data.csv')

# Review the DataFrame
crdf.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [78]:
# Separate the data into labels and features

# Separate the y variable, the labels
y=crdf['loan_status']

# Separate the X variable, the features
X=crdf.drop(columns='loan_status')

In [79]:
# Review the y variable Series
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [80]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [81]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [82]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [83]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [84]:
# Make a prediction using the testing data
predicted_y_testvalues = logistic_regression_model.predict(X_test)
predicted_y_testvalues

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [85]:
# Print the balanced_accuracy score of the model
balanced_accuracy_score(y_test,predicted_y_testvalues)

0.9520479254722232

In [86]:
# Generate a confusion matrix for the model
confusion_matrix(y_test,predicted_y_testvalues)

array([[18663,   102],
       [   56,   563]], dtype=int64)

In [87]:
# Print the classification report for the model
print(classification_report(y_test,predicted_y_testvalues))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
#consulted ChatGPT to interprete results

A balanced score of 0.9520479254722232 indicates that the model has a high average recall across all classes, suggesting that it is able to predict both the both the `0` (healthy loan) and `1` (high-risk loan) labels accurately

Precision: Precision measures the proportion of positive predictions that were correct. In this case, for class 0, the precision is 1.00, which means that all of the samples predicted as 0 were actually 0. For class 1, the precision is 0.85, which means that 85% of the samples predicted as 1 were actually 1.

Recall: Recall measures the proportion of actual positive samples that were correctly predicted. In this case, for class 0, the recall is 0.99, which means that 99% of the actual 0 samples were correctly predicted as 0. For class 1, the recall is 0.91, which means that 91% of the actual 1 samples were correctly predicted as 1.

F1-score: F1-score is the harmonic mean of precision and recall. It provides a single score that balances both precision and recall. In this case, for class 0, the F1-score is 1.00, which indicates that the model is performing well for class 0. For class 1, the F1-score is 0.88, which indicates that there is some room for improvement in the model's performance for class 1.

Overall, the classification report indicates that the model is performing well for class 0 with high precision, recall, and F1-score. However, the model's performance for class 1 is not as good, with lower precision, recall, and F1-score.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [88]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
oversampler=RandomOverSampler(random_state=1)


# Fit the original training data to the random_oversampler model
X_over, y_over = oversampler.fit_resample(X_test, y_test)

In [89]:
# Count the distinct values of the resampled labels data
y_over.value_counts()

0    18765
1    18765
Name: loan_status, dtype: int64

In [91]:
#!pip install imblearn

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [92]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(random_state=1)


# Fit the model using the resampled training data
lr_model = logistic_regression_model.fit(X_over, y_over)

# Make a prediction using the testing data
predicted_y_values = logistic_regression_model.predict(X_test)
predicted_y_values

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [93]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(y_test,predicted_y_values)

0.9936781215845847

In [94]:
# Generate a confusion matrix for the model
confusion_matrix(y_test,predicted_y_values)

array([[18649,   116],
       [    4,   615]], dtype=int64)

In [95]:
# Print the classification report for the model
print(classification_report(y_test, predicted_y_values))

              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
#consulted ChatGPT to interprete results
A balanced accuracy score of 0.9936781215845847 indicates that the model has a very high average sensitivity (recall) across all classes, suggesting that it is able to predict both the majority and minority classes accurately

Precision: For class 0, the precision is 1.00, which means that all of the samples predicted as 0 were actually 0. For class 1, the precision is 0.84, which means that 84% of the samples predicted as 1 were actually 1.

Recall: Recall measures the proportion of actual positive samples that were correctly predicted. In this case, for class 0, the recall is 0.99, which means that 99% of the actual 0 samples were correctly predicted as 0. For class 1, the recall is 0.99, which means that 99% of the actual 1 samples were correctly predicted as 1.

F1-score: F1-score is the harmonic mean of precision and recall. It provides a single score that balances both precision and recall. In this case, for class 0, the F1-score is 1.00, which indicates that the model is performing well for class 0. For class 1, the F1-score is 0.91, which indicates that there is some room for improvement in the model's performance for class 1.

Overall, the classification report indicates that the model is performing well for both classes with high precision, recall, and F1-score. However, the model's performance for class 1 is slightly lower than that for class 0, with a lower precision and F1-score.